tmt/ansible.py

```diff
@@ -0,0 +1,336 @@
+
+        raise tmt.utils.SpecificationError(f"Invalid Ansible specification: {spec}")
+
+    def to_spec(self) -> dict[str, Any]:
+        """
+        Convert PlanAnsible object to a YAML-serializable specification.
+        """
+        spec: dict[str, Any] = {}
+
+        if self.inventory is not None:
+            spec['inventory'] = self.inventory.to_spec()
+
+        return spec
+
+    def __repr__(self) -> str:
+        """Return a string representation of the PlanAnsible object."""
+        return f"PlanAnsible(inventory={self.inventory})"
+
+
+class AnsibleInventory:
+    """
+    Generate Ansible inventory files from provisioned guests.
+
+    Creates Ansible inventory file that can be used with playbooks
+    to manage provisioned guests. Supports custom layouts and automatically
+    configures host variables based on guest properties.
+    """
+
+    def __init__(self, logger: tmt.log.Logger) -> None:
+        """
+        Initialize the Ansible inventory handler.
+
+        :param logger: logger instance for debug/info messages.
+        """
+        self._logger = logger
+
+    def _load_layout(self, layout_path: Optional[Path] = None) -> dict[str, Any]:
+        """
+        Load inventory layout from file or use default, ensuring required Ansible structure.
+
+        :param layout_path: full path to a custom layout file.
+        :returns: dictionary representing the inventory layout structure with required groups.
+        """
+        if layout_path:
+            try:
+                layout = tmt.utils.yaml_to_dict(layout_path.read_text())
+            except (OSError, FileNotFoundError):
+                raise tmt.utils.FileError(f"Inventory layout file '{layout_path}' not found")
+        else:
+            layout = {}
+
+        return self._normalize_layout(layout)
+
+    def _normalize_layout(self, layout: dict[str, Any]) -> dict[str, Any]:
+        """
+        Normalize layout to ensure required Ansible inventory structure.
+
+        Ensures 'all' and 'ungrouped' groups are present.
+
+        :param layout: raw layout dictionary from file or empty dict.
+        :returns: normalized inventory with required Ansible structure.
+        """
+        # Ensure 'all' group exists
+        if 'all' not in layout:
+            layout['all'] = {}
+
+        # Ensure 'ungrouped' exists for hosts without explicit groups
+        if 'children' not in layout['all']:
+            layout['all']['children'] = {}
+        if 'ungrouped' not in layout['all']['children']:
+            layout['all']['children']['ungrouped'] = {}
+
+        return layout
+
+    def _add_host_to_all(self, inventory: dict[str, Any], guest: 'Guest') -> None:
+        """
+        Add host with its variables to the 'all' group.
+
+        :param inventory: the inventory dictionary to modify.
+        :param guest: the guest to add to the inventory.
+        """
+        host_vars = guest.ansible_host_vars
```

Comment:
`host_vars` exists only to be assigned to `inventory['all']['hosts'][guest.name]`, not needed.